stakedb/upgrade.go

```diff
@@ -0,0 +1,291 @@
+package stakedb
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+
+	bv1 "github.com/dgraph-io/badger"
+	"github.com/dgraph-io/badger/v3"
+)
+
+type upgradeFunc func(string) error
+
+// Each database upgrade function should be keyed by the database
+// version it upgrades.
+var upgrades = [...]upgradeFunc{
+	// v0 => v1, rewrite db using custom encoding.
+	v1Upgrade,
+	// v1 => v2, upgrades db to new badger v3.
+	v2Upgrade,
+}
+
+func v1DBVersion(db *bv1.DB) (uint32, error) {
+	var ver uint32
+	return ver, db.View(func(txn *bv1.Txn) error {
+		verItem, err := txn.Get(versionKey)
+		if err != nil {
+			if errors.Is(err, bv1.ErrKeyNotFound) {
+				log.Debug("Detected v0 pool db with no version key.")
+				return nil // ver = 0
+			}
+			return err
+		}
+
+		if sz := verItem.ValueSize(); sz != 4 {
+			return fmt.Errorf("invalid version value size %d", sz)
+		}
+
+		return verItem.Value(func(v []byte) error {
+			ver = binary.BigEndian.Uint32(v) // should not be 0
+			return nil
+		})
+	})
+}
+
+// canUpgrade checks if a v1 badger database can upgrade.
+func canUpgrade(db *bv1.DB, expectedVer uint32) (bool, error) {
+	ver, err := v1DBVersion(db)
+	if err != nil {
+		return false, fmt.Errorf("canUpgrade error: %w", err)
+	}
+
+	if ver > currentVersion {
+		return false, fmt.Errorf("unknown database version %d, "+
+			"ticket pool db recognizes up to %d", ver, currentVersion)
+	}
+
+	if ver == expectedVer {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// openV1DB opens a v1 badger database.
+func openV1DB(dbPath string) (*bv1.DB, error) {
+	opts := bv1.DefaultOptions(dbPath).
+		WithValueDir(dbPath).
+		WithNumVersionsToKeep(math.MaxUint32).
```

Comment:
Why would we do this? The default of 1 version per key to keep is appropriate, I believe, especially now that we're only using v1 badger DBs in a read-only capacity (unless we need to truncate to recover from corruption).